File: backend/src/ws/handler.rs
use std::arch::global_asm;

use axum::extract::ws::{Message, WebSocket};
use futures_util::{SinkExt, StreamExt};
use tokio::sync::mpsc;
use tracing::{error, info};

use crate::game::actor::run_game_actor;
use crate::game::command::GameCommand;
use crate::lobby::command::LobbyCommand;
use crate::room::command::RoomCommand;
use crate::room::room::ClientDataForState;
use crate::state::{get_global_state, AppState, ClientState};
use crate::ws::message::{
    make_ws_msg_connected, ConnectedPayload, GenerateBombPayload, PlayerMovePayload,
    TimeSyncPingPayload, WsRawMessage,
};

pub async fn handle_init_socket(socket: WebSocket) {
    let state = get_global_state();
    // let client_id = state.alloc_client_id();
    let client_id =1;
    let (sender, mut receiver) = mpsc::unbounded_channel::<Message>();
    state.client_sender_map.insert(client_id, sender.clone());
    state.client_state_map.insert(
        client_id,
        ClientState {
            id: client_id,
            name: String::new(),
            room_id: None,
            game_id: Some(1),
        },
    );

    // Notify the lobby actor that a new client is waiting.
    let _ = state
        .lobby_sender
        .send(LobbyCommand::ClientJoined(client_id));
    info!("client {} connected", client_id);

    let (mut ws_sender, mut ws_receiver) = socket.split();
    // Dedicated task: forward mpsc messages to the WebSocket.
    let send_task = tokio::spawn(async move {
        while let Some(msg) = receiver.recv().await {
            if ws_sender.send(msg).await.is_err() {
                break;
            }
        }
    });

    // Tell the client their assigned ID.
    state.send_to_client(
        client_id,
        Message::Text(make_ws_msg_connected(ConnectedPayload { user_id: client_id }).into()),
    );
    test_game_start(client_id);
    // let _ = sender.send(Message::Text(
    //     make_ws_msg("startGame", &ConnectedPayload { user_id: client_id }).into(),
    // ));

    // Receive loop.
    while let Some(Ok(msg)) = ws_receiver.next().await {
        match msg {
            Message::Text(text) => {
                if let Err(e) = handle_msg(client_id, &text, state).await {
                    error!("handle_msg error for client {}: {}", client_id, e);
                }
            }
            Message::Close(_) => break,
            _ => {}
        }
    }

    info!("client {} disconnected", client_id);
    cleanup_client(client_id, state).await;
    send_task.await.ok();
}

async fn cleanup_client(client_id: u32, state: &'static AppState) {
    let (room_id, game_id) = state
        .client_state_map
        .get(&client_id)
        .map(|cs| (cs.room_id, cs.game_id))
        .unwrap_or((None, None));

    if let Some(rid) = room_id {
        state.send_to_room(rid, RoomCommand::PlayerLeave { client_id });
    } else if let Some(gid) = game_id {
        state.send_to_game(gid, GameCommand::PlayerDisconnected(client_id));
    } else {
        // Client was still in the lobby.
        let _ = state.lobby_sender.send(LobbyCommand::ClientLeft(client_id));
    }

    state.client_sender_map.remove(&client_id);
    state.client_state_map.remove(&client_id);
}

async fn handle_msg(
    client_id: u32,
    text: &str,
    state: &'static AppState,
) -> Result<(), serde_json::Error> {
    let raw: WsRawMessage = serde_json::from_str(text)?;
    match raw.msg_type.as_str() {
        "setName" => {
            let name: String = serde_json::from_value(raw.payload)?;
            if let Some(mut cs) = state.client_state_map.get_mut(&client_id) {
                cs.name = name;
            }
        }

        "createRoom" => {
            let name = state
                .client_state_map
                .get(&client_id)
                .map(|cs| cs.name.clone())
                .unwrap_or_default();
            let room_id = state.alloc_room_id();
            let (room_tx, room_rx) = mpsc::unbounded_channel::<RoomCommand>();
            state.room_sender_map.insert(room_id, room_tx);
            if let Some(mut cs) = state.client_state_map.get_mut(&client_id) {
                cs.room_id = Some(room_id);
            }
            let _ = state.lobby_sender.send(LobbyCommand::ClientLeft(client_id));
            tokio::spawn(crate::room::actor::run_room_actor(
                room_id, client_id, name, room_rx, state,
            ));
        }

        "joinRoom" => {
            let room_id: u32 = serde_json::from_value(raw.payload)?;
            let name = state
                .client_state_map
                .get(&client_id)
                .map(|cs| cs.name.clone())
                .unwrap_or_default();
            state.send_to_room(room_id, RoomCommand::PlayerJoin { client_id, client_name: name });
        }

        "leaveRoom" => {
            if let Some(rid) = state.client_state_map.get(&client_id).and_then(|cs| cs.room_id) {
                state.send_to_room(rid, RoomCommand::PlayerLeave { client_id });
            }
        }

        "toggleReady" => {
            if let Some(rid) = state.client_state_map.get(&client_id).and_then(|cs| cs.room_id) {
                state.send_to_room(rid, RoomCommand::ToggleReady { client_id });
            }
        }

        "changeMap" => {
            let map_id: u32 = serde_json::from_value(raw.payload)?;
            if let Some(rid) = state.client_state_map.get(&client_id).and_then(|cs| cs.room_id) {
                state.send_to_room(rid, RoomCommand::ChangeMap { map_id });
            }
        }

        "startGame" => {
            if let Some(rid) = state.client_state_map.get(&client_id).and_then(|cs| cs.room_id) {
                state.send_to_room(rid, RoomCommand::StartGame { host_client_id: client_id });
            }
        }

        "playerMove" => {
            let payload: PlayerMovePayload = serde_json::from_value(raw.payload)?;
            println!("{}",5);
            if let Some(gid) = state.client_state_map.get(&client_id).and_then(|cs| cs.game_id) {
                state.send_to_game(gid, GameCommand::PlayerMove(payload));
            }
        }

        "generateBomb" => {
            let payload: GenerateBombPayload = serde_json::from_value(raw.payload)?;
            if let Some(gid) = state.client_state_map.get(&client_id).and_then(|cs| cs.game_id) {
                state.send_to_game(gid, GameCommand::GenerateBomb(payload));
            }
        }

        "timeSyncPing" => {
            let payload: TimeSyncPingPayload = serde_json::from_value(raw.payload)?;
            if let Some(gid) = state.client_state_map.get(&client_id).and_then(|cs| cs.game_id) {
                state.send_to_game(gid, GameCommand::TimeSyncPing { client_id, sent_at: payload.sent_at });
            }
        }

        other => {
            info!("unknown message type '{}' from client {}", other, client_id);
        }
    }

    Ok(())
}

pub fn test_game_start(host_client_id: u32) {
    let global_state = get_global_state();
    let game_id = 1;
    let (game_tx, game_rx) = tokio::sync::mpsc::unbounded_channel::<GameCommand>();
    global_state.game_sender_map.insert(game_id, game_tx);

    let mut players: Vec<ClientDataForState> = Vec::new();
    players.push(ClientDataForState {
        id: host_client_id,
        name: "test".to_string(),
        is_host: true,
        is_ready: true,
        man_sprite_key: crate::game::types::ManSpriteKey::Man1,
    });
    tokio::spawn(run_game_actor(
        game_id,
        players,
        1,
        game_rx,
        global_state,
    ));
    // info!("room {} started game {}", room_id, game_id);
}

File: backend2/ws/handler.go
package ws

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"

	"github.com/gorilla/websocket"

	"bomberman/backend2/game"
	"bomberman/backend2/room"
	"bomberman/backend2/state"
	"bomberman/backend2/types"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

type wsRawMessage struct {
	MsgType string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

func makeWsMsg(msgType string, payload interface{}) []byte {
	b, _ := json.Marshal(map[string]interface{}{
		"type":    msgType,
		"payload": payload,
	})
	return b
}

// HandleWs upgrades the HTTP connection and runs the client loop.
func HandleWs(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("ws upgrade error: %v", err)
		return
	}

	appState := state.GetGlobalState()
	clientID := appState.AllocClientID()

	sendCh := make(chan []byte, 256)
	appState.ClientSenderMap.Store(clientID, sendCh)
	appState.ClientStateMap.Store(clientID, &state.ClientState{ID: clientID})

	appState.LobbySender <- state.LobbyCommand{Kind: state.LobbyClientJoined, ClientID: clientID}
	log.Printf("client %d connected", clientID)

	// Send assigned ID
	appState.SendToClient(clientID, makeWsMsg("connected", map[string]interface{}{"userId": clientID}))

	// Mirror Rust test_game_start: auto-start a debug game
	testGameStart(clientID, appState)

	// Write pump
	go func() {
		defer conn.Close()
		for msg := range sendCh {
			if err := conn.WriteMessage(websocket.TextMessage, msg); err != nil {
				break
			}
		}
	}()

	// Read pump (blocks until disconnect)
	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			break
		}
		var msg wsRawMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			log.Printf("client %d bad message: %v", clientID, err)
			continue
		}
		if err := handleMsg(clientID, msg, appState); err != nil {
			log.Printf("client %d handle_msg error: %v", clientID, err)
		}
	}

	log.Printf("client %d disconnected", clientID)
	cleanupClient(clientID, appState)
	close(sendCh)
}

func cleanupClient(clientID uint32, appState *state.AppState) {
	cs := appState.GetClientState(clientID)
	var roomID, gameID *uint32
	if cs != nil {
		roomID, gameID = cs.GetRoomAndGameID()
	}
	switch {
	case roomID != nil:
		appState.SendToRoom(*roomID, state.RoomCommand{Kind: state.RoomPlayerLeave, ClientID: clientID})
	case gameID != nil:
		appState.SendToGame(*gameID, state.GameCommand{Kind: state.GamePlayerDisconnected, ClientID: clientID})
	default:
		appState.LobbySender <- state.LobbyCommand{Kind: state.LobbyClientLeft, ClientID: clientID}
	}
	appState.ClientSenderMap.Delete(clientID)
	appState.ClientStateMap.Delete(clientID)
}

func handleMsg(clientID uint32, msg wsRawMessage, appState *state.AppState) error {
	cs := appState.GetClientState(clientID)
	fmt.Print(msg)
	switch msg.MsgType {
	case "setName":
		var name string
		if err := json.Unmarshal(msg.Payload, &name); err != nil {
			return err
		}
		if cs != nil {
			cs.SetName(name)
		}

	case "createRoom":
		name := ""
		if cs != nil {
			name = cs.GetName()
		}
		roomID := appState.AllocRoomID()
		roomCh := make(chan state.RoomCommand, 256)
		appState.RoomSenderMap.Store(roomID, roomCh)
		if cs != nil {
			rid := roomID
			cs.SetRoomID(&rid)
		}
		appState.LobbySender <- state.LobbyCommand{Kind: state.LobbyClientLeft, ClientID: clientID}
		go room.RunRoomActor(roomID, clientID, name, roomCh, appState)

	case "joinRoom":
		var roomID uint32
		if err := json.Unmarshal(msg.Payload, &roomID); err != nil {
			return err
		}
		name := ""
		if cs != nil {
			name = cs.GetName()
		}
		appState.SendToRoom(roomID, state.RoomCommand{Kind: state.RoomPlayerJoin, ClientID: clientID, ClientName: name})

	case "leaveRoom":
		if cs != nil {
			if rid := cs.GetRoomID(); rid != nil {
				appState.SendToRoom(*rid, state.RoomCommand{Kind: state.RoomPlayerLeave, ClientID: clientID})
			}
		}

	case "toggleReady":
		if cs != nil {
			if rid := cs.GetRoomID(); rid != nil {
				appState.SendToRoom(*rid, state.RoomCommand{Kind: state.RoomToggleReady, ClientID: clientID})
			}
		}

	case "changeMap":
		var mapID uint32
		if err := json.Unmarshal(msg.Payload, &mapID); err != nil {
			return err
		}
		if cs != nil {
			if rid := cs.GetRoomID(); rid != nil {
				appState.SendToRoom(*rid, state.RoomCommand{Kind: state.RoomChangeMap, MapID: mapID})
			}
		}

	case "startGame":
		if cs != nil {
			if rid := cs.GetRoomID(); rid != nil {
				appState.SendToRoom(*rid, state.RoomCommand{Kind: state.RoomStartGame, HostClientID: clientID})
			}
		}

	case "playerMove":
		var p state.PlayerMovePayload
		if err := json.Unmarshal(msg.Payload, &p); err != nil {
			return err
		}
		if cs != nil {
			if gid := cs.GetGameID(); gid != nil {
				appState.SendToGame(*gid, state.GameCommand{Kind: state.GamePlayerMove, Move: &p})
			}
		}

	case "generateBomb":
		var p state.GenerateBombPayload
		if err := json.Unmarshal(msg.Payload, &p); err != nil {
			return err
		}
		if cs != nil {
			if gid := cs.GetGameID(); gid != nil {
				appState.SendToGame(*gid, state.GameCommand{Kind: state.GameGenerateBomb, Bomb: &p})
			}
		}

	case "timeSyncPing":
		var p struct {
			SentAt int64  `json:"sentAt"`
			From   string `json:"from"`
		}
		if err := json.Unmarshal(msg.Payload, &p); err != nil {
			return err
		}
		if cs != nil {
			if gid := cs.GetGameID(); gid != nil {
				appState.SendToGame(*gid, state.GameCommand{Kind: state.GameTimeSyncPing, ClientID: clientID, SentAt: p.SentAt})
			}
		}

	default:
		log.Printf("unknown message type '%s' from client %d", msg.MsgType, clientID)
	}
	return nil
}

func testGameStart(hostClientID uint32, appState *state.AppState) {
	gameID := uint32(1)
	gameCh := make(chan state.GameCommand, 256)
	appState.GameSenderMap.Store(gameID, gameCh)
	if cs := appState.GetClientState(hostClientID); cs != nil {
		gid := gameID
		cs.SetGameID(&gid)
	}
	players := []types.PlayerInfo{{
		ID:           hostClientID,
		Name:         "test",
		IsReady:      true,
		IsHost:       true,
		ManSpriteKey: types.Man1,
	}}
	go game.RunGameActor(gameID, players, 1, gameCh, appState)
}
